presto-generation-processor/src/main/java/com/facebook/presto/test/ConnectorTestProcessor.java

```diff
@@ -0,0 +1,244 @@
+                .map(method -> method.getSimpleName().toString())
+                .collect(toImmutableList());
+    }
+
+    private void addTests(Set<TypeElement> newTests)
+    {
+        generate(connectors, newTests);
+
+        tests = ImmutableSet.<TypeElement>builder()
+                .addAll(tests)
+                .addAll(newTests)
+                .build();
+    }
+
+    private void addConnectors(Set<TypeElement> newConnectors)
+    {
+        generate(newConnectors, tests);
+
+        connectors = ImmutableSet.<TypeElement>builder()
+                .addAll(connectors)
+                .addAll(newConnectors)
+                .build();
+    }
+
+    private static Set<ConnectorFeature> supportedFeatures(TypeElement connector)
+    {
+        return ImmutableSet.copyOf(connector.getAnnotation(TestableConnector.class).supportedFeatures());
+    }
+
+    private static Set<ConnectorFeature> requiredFeatures(Element test)
+    {
+        return ImmutableSet.copyOf(test.getAnnotation(ConnectorTest.class).requiredFeatures());
+    }
+
+    private boolean connectorSupportsTest(TypeElement connector, TypeElement test)
+    {
+        Set<ConnectorFeature> supportedFeatures = supportedFeatures(connector);
+        Set<ConnectorFeature> requiredFeatures = requiredFeatures(test);
+
+        return supportedFeatures.containsAll(requiredFeatures);
+    }
+
+    private boolean connectorSupportsTestMethod(TypeElement connector, TypeElement test, ExecutableElement testMethod)
+    {
+        Set<ConnectorFeature> supportedFeatures = supportedFeatures(connector);
+        Set<ConnectorFeature> requiredFeatures = Sets.union(requiredFeatures(test), requiredFeatures(testMethod));
+
+        return supportedFeatures.containsAll(requiredFeatures);
+    }
+
+    private static String getTestClassName(TypeElement connector, TypeElement test)
+    {
+        String connectorName = connector.getAnnotation(TestableConnector.class).connectorName();
+        String testName = test.getAnnotation(ConnectorTest.class).testName();
+        return "Test" + testName + connectorName;
+    }
+
+    private static String getPackageName(Element element)
+    {
+        if (PackageElement.class.isAssignableFrom(element.getClass())) {
+            return ((PackageElement) element).getQualifiedName().toString();
+        }
+
+        return getPackageName(element.getEnclosingElement());
+    }
+
+    private static String getFqClassName(TypeElement connector, TypeElement test)
+    {
+        return getPackageName(connector) + "." + getTestClassName(connector, test);
+    }
+
+    private void generate(Set<TypeElement> connectors, Set<TypeElement> tests)
+    {
+        for (TypeElement connector : connectors) {
+            tests.stream()
+                    .filter(test -> connectorSupportsTest(connector, test))
+                    .forEach(test -> generate(connector, test));
+        }
+    }
+
+    private void generate(TypeElement connector, TypeElement test)
```

Comment:
Here and thoughout, the fact that connectors and tests are of the same type (TypeElement) creates ample opportunities to pass the wrong thing. If this gets much more complex, we should wrap them in a more specific type to make troubleshooting easier. Or if people want that now, it can be done.